### PR TITLE
Support early print buffer flushing in debug mode

### PR DESCRIPTION
**Pull request by zibaiwan**
Added a flow to flush printf() buffer before the Kernel ends. This will be helpful for debugging a hang Kernel. The Runtime periodically polls from the device and process the printf buffer from the last processed point.

If ACL_HAL_DEBUG = 1 and the Kernel is Hang, Kernel status (existing feature) and printf buffer will be flushed to stdout every 10 seconds (approximately). The thread will run if there are spare cycles.

If ACL_HAL_DEBUG = 3, Kernel status and printf buffer will flushed every 10 seconds, even the Kernel isn't hung.

The behavior of printf() is not altered in non-debug mode.

## Sample output

```
  Kernel  0 Status: 0x00048000 running:: Accelerator 0 printf buffer size is 160.
:: Calling acl_process_printf_buffer_fn with activation_id=1 and printf_size=160.


Previously processed buffer size is 0 
Another test string 14
This is a test
Another test string 11
Another test string 12
Another test string 13
  Kernel  0 Status: 0x00048000 running:: Accelerator 0 printf buffer size is 160.
:: Calling acl_process_printf_buffer_fn with activation_id=1 and printf_size=160.


Previously processed buffer size is 160 
All Printf() buffer has already been dumped
```

## ToDo
- [x] Fix Klocwork if applicable
- [x] Run Clang format
- [x] manual testing
- [x] Clear commit messages and Rebase
- [x] Run existing regtest to ensure existing feature still passes
- [ ] Create a new regtest to test this specific feature

### CODE REVIEW

File: src/acl_printf.cpp [hunk 1]
```diff
@@ -769,8 +770,24 @@ static void l_dump_printf_buffer(cl_event event, cl_kernel kernel,
     hal_dma_fn = acl_get_hal()->copy_globalmem_to_hostmem;
   }
 
-  hal_dma_fn(NULL, kernel->printf_device_buffer->block_allocation->range.begin,
-             buffer, size);
+  // It needs the context from ACL_HAL_DEBUG instead of ACL_DEBUG
+  if (acl_get_hal()->get_debug_verbosity &&
+      acl_get_hal()->get_debug_verbosity() > 0) {
+    printf("Previously processed buffer size is %d \n", kernel->processed_printf_buffer_size);
```
Comment (pcolberg) [inline src/acl_printf.cpp:776]: For `size_t` the printf modifier is `%zu`.

File: include/acl_types.h [hunk 2]
```diff
@@ -1346,6 +1350,10 @@ typedef struct {
   // it sees the CL_COMPLETE status for the first time, but it won't
   // change after that.
   cl_uint num_printf_bytes_pending;
+
+  // Indicate whether this operation is dumping printf buffer before the Kernel
+  // for debug purpose
+  int debug_dump_printf = 0;
```
Comment (pcolberg) [inline include/acl_types.h:1356]: Would anything speak against making this a `bool`, since it only has two states?

File: src/acl_printf.cpp [hunk 3]
```diff
@@ -769,8 +770,28 @@ static void l_dump_printf_buffer(cl_event event, cl_kernel kernel,
...
+           kernel->processed_printf_buffer_size);
+  }
+
+  // Check if we have already processed all the printf buffer
+  if (size > (unsigned int)kernel->processed_printf_buffer_size) {
+    void *unprocessed_begin = (void *)((char *)kernel->printf_device_buffer
+                                           ->block_allocation->range.begin +
+                                       kernel->processed_printf_buffer_size);
+    hal_dma_fn(NULL, unprocessed_begin, buffer,
+               size - kernel->processed_printf_buffer_size);
```
Comment (pcolberg) [inline src/acl_printf.cpp:786]: How about moving the assignment to `dumped_buffer_size` before this call and using it here?

Comment (pcolberg) [reply src/acl_printf.cpp:786]: `assert(size >= kernel->processed_printf_buffer_size)` might be useful to catch a possible underflow.

File: src/acl_printf.cpp [hunk 4]
```diff
@@ -960,13 +981,14 @@ static void l_dump_printf_buffer(cl_event event, cl_kernel kernel,
...
 #endif
+  return dumped_buffer_size;
 }
 
 //
 // Schedule enqueue read buffer to read printf buffer
 // The activation ID is the device op ID.
 void acl_schedule_printf_buffer_pickup(int activation_id, int size,
-                                       int stalled) {
+                                       int debug_dump_printf) {
```
Comment (pcolberg) [inline src/acl_printf.cpp:991]: Does this HAL API also need to be updated?
https://github.com/intel/fpga-runtime-for-opencl/blob/67f7f880f7af5af0c1797c4dd2a70b4e27ba85ca/include/acl_hal.h#L65-L66

File: include/acl_types.h (hunk 2 above)
Comment (pcolberg) [reply include/acl_types.h:1356]: Never mind, the type `int` is aligned with the HAL API, correct?
https://github.com/intel/fpga-runtime-for-opencl/blob/67f7f880f7af5af0c1797c4dd2a70b4e27ba85ca/include/acl_hal.h#L65-L66

File: src/acl_printf.cpp [hunk 5]
```diff
@@ -980,17 +1002,21 @@ void acl_schedule_printf_buffer_pickup(int activation_id, int size,
...
     // This address is stable, given a fixed activation_id.
     // So we don't run into race conditions.
     acl_device_op_t *op = doq->op + activation_id;
-
-    stalled = stalled; // this argument is no longer used!
-
     op->info.num_printf_bytes_pending = (cl_uint)size;
+
+    // Propagate the operation info
+    if (debug_dump_printf == 1) {
```
Comment (pcolberg) [inline src/acl_printf.cpp:1015]: What happens if `debug_dump_printf` is any other non-zero value, say 2?

File: src/acl_printf.cpp [hunk 6]
```diff
@@ -980,17 +1002,21 @@ void acl_schedule_printf_buffer_pickup(int activation_id, int size,
...
+      op->info.debug_dump_printf = 1;
+    } else {
+      op->info.debug_dump_printf = 0;
+    }
```
Comment (pcolberg) [inline src/acl_printf.cpp:1019]: You could shorten to
```
op->info.debug_dump_printf = debug_dump_printf ? 1 : 0;
```
i.e., zero input value zero yields 0 and non-zero input value is normalized to 1.

File: src/acl_printf.cpp [hunk 7]
```diff
@@ -980,17 +1002,21 @@ void acl_schedule_printf_buffer_pickup(int activation_id, int size,
...
   fflush(stdout);
 #endif
   if (activation_id >= 0 && activation_id < doq->max_ops) {
     // This address is stable, given a fixed activation_id.
     // So we don't run into race conditions.
     acl_device_op_t *op = doq->op + activation_id;
-
-    stalled = stalled; // this argument is no longer used!
-
     op->info.num_printf_bytes_pending = (cl_uint)size;
```
Comment (pcolberg) [inline src/acl_printf.cpp:1012]: Why is `num_printf_bytes_pending` not `size_t`?

**Review by pcolberg — CHANGES_REQUESTED** [review]
Thanks @zibaiwan! Could you include a commit message that summarizes the change and also motivates why it was made?

Have you taken a look at the unit tests to see if anything could be inserted to test this new functionality, as far as is feasible in a unit test? For instance, checking that the new state variables added here correspond to the expected state.

File: src/acl_printf.cpp (hunk 7 above)
Comment (pcolberg) [reply src/acl_printf.cpp:1012]: Please never mind, this is not part of the change.

File: src/acl_printf.cpp (hunk 6 above)
Comment (zibaiwan) [reply src/acl_printf.cpp:1019]: Thanks Peter. Your suggestion looks cleaner and handled any other non-zero value.

**Comment by zibaiwan** [comment]
@pcolberg , thank you very much for your review. I have updated the PR based on your suggestions.

File: src/acl_printf.cpp [hunk 8]
```diff
@@ -1009,7 +1031,15 @@ void acl_process_printf_buffer(void *user_data, acl_device_op_t *op) {
 
     // Grab the printf data and emit it.
     cl_uint num_bytes = op->info.num_printf_bytes_pending;
-    l_dump_printf_buffer(event, kernel, num_bytes);
+    size_t dumped_buffer_size = l_dump_printf_buffer(event, kernel, num_bytes);
+
+    if (op->info.debug_dump_printf == 1) {
```
Comment (pcolberg) [inline src/acl_printf.cpp:1036]: Since `int debug_dump_printf` has boolean semantics:
```c++
     if (op->info.debug_dump_printf) {
```

File: src/acl_printf.cpp [hunk 9]
```diff
@@ -1021,8 +1051,13 @@ void acl_process_printf_buffer(void *user_data, acl_device_op_t *op) {
     acl_memory_barrier();
 
     // Allow the kernel to continue running.
-    acl_get_hal()->unstall_kernel(
-        event->cmd.info.ndrange_kernel.device->def.physical_device_id, op->id);
+    // We don't need to unstall the kernel during the early flushing during
+    // debug.
+    if (op->info.debug_dump_printf == 0) {
```
Comment (pcolberg) [inline src/acl_printf.cpp:1056]:
```c++
    if (!op->info.debug_dump_printf) {
```

**Review by pcolberg — CHANGES_REQUESTED** [review]
Thanks @zibaiwan for the additional tests, much appreciated :slightly_smiling_face:

**Comment by sherry-yuan** [comment]
A small question about a (maybe) corner case: What happens if the kernel stalls before the next 10 sec printf buffer dumps?

**Comment by zibaiwan** [comment]
> A small question about a (maybe) corner case: What happens if the kernel stalls before the next 10 sec printf buffer dumps?

Good question. If a kernel is stalled because the printf buffer is full (runtime doesn't seem to stall for other reason, even it does, it wouldn't matter here), it will print the rest of the printf buffer which hasn't been dumped by the debug dumps.

File: src/acl_printf.cpp (hunk 9 above)
Comment (zibaiwan) [reply src/acl_printf.cpp:1056]: Thanks for the suggestion!

**Review by pcolberg — APPROVED** [review]
